File: apps/api/src/models/ModelQuery.ts
import {
  Model,
  DataTypes,
  Optional,
  BelongsToGetAssociationMixin,
  HasOneGetAssociationMixin,
} from "sequelize";
import type { Agent } from "./Agent";
import type { HumanTask } from "./HumanTask";
import type { PerformanceMetric } from "./PerformanceMetric";
import { sequelize } from "../config/database";

export type QueryMethod = "PERFORMANCE" | "EFFICIENCY";

export interface ModelQueryAttributes {
  id: string;
  agentId: string;
  userQuery: string;
  modelResponse?: string | null;
  methodUsed: QueryMethod;
  createdAt?: Date;
  updatedAt?: Date;
  isDeleted?: boolean;
  deletedAt?: Date | null;
}

export type ModelQueryCreationAttributes = Optional<
  ModelQueryAttributes,
  "id" | "modelResponse" | "createdAt" | "updatedAt" | "isDeleted" | "deletedAt"
>;

export class ModelQuery
  extends Model<ModelQueryAttributes, ModelQueryCreationAttributes>
  implements ModelQueryAttributes
{
  declare id: string;
  declare agentId: string;
  declare userQuery: string;
  declare modelResponse: string | null;
  declare methodUsed: QueryMethod;
  declare readonly createdAt: Date;
  declare readonly updatedAt: Date;
  declare isDeleted: boolean;
  declare deletedAt: Date | null;

  declare getAgent: BelongsToGetAssociationMixin<Agent>;
  declare getHumanTask: HasOneGetAssociationMixin<HumanTask>;
  declare getMetrics: HasOneGetAssociationMixin<PerformanceMetric>;
}

ModelQuery.init(
  {
    id: {
      type: DataTypes.UUID,
      defaultValue: DataTypes.UUIDV4,
      primaryKey: true,
    },
    agentId: {
      type: DataTypes.UUID,
      allowNull: false,
      field: "agent_id",
    },
    userQuery: {
      type: DataTypes.TEXT,
      allowNull: false,
      field: "user_query",
    },
    modelResponse: {
      type: DataTypes.TEXT,
      allowNull: true,
      field: "model_response",
    },
    methodUsed: {
      type: DataTypes.ENUM("PERFORMANCE", "EFFICIENCY"),
      allowNull: false,
      field: "method_used",
    },
    isDeleted: {
      type: DataTypes.BOOLEAN,
      defaultValue: false,
      field: "is_deleted",
    },
    deletedAt: {
      type: DataTypes.DATE,
      allowNull: true,
      field: "deleted_at",
    },
  },
  {
    sequelize,
    tableName: "model_queries",
    indexes: [{ fields: ["agent_id"] }],
    defaultScope: { where: { isDeleted: false } },
  },
);

File: python/app/models/model_query.py
"""ModelQuery model: per-agent query log."""
# pyright: ignore[reportUndefinedVariable]

from __future__ import annotations

import uuid

from sqlalchemy import Boolean, DateTime, ForeignKey, String, Text
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship
from sqlalchemy.sql import func

from app.models.base import Base


class ModelQuery(Base):
    __tablename__ = "model_queries"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    agent_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("agents.id", onupdate="CASCADE", ondelete="RESTRICT"), nullable=False, index=True
    )
    user_query: Mapped[str] = mapped_column(Text(), nullable=False)
    model_response: Mapped[str | None] = mapped_column(Text(), nullable=True)
    method_used: Mapped[str] = mapped_column(String(), nullable=False)  # PERFORMANCE, EFFICIENCY
    created_at: Mapped[DateTime] = mapped_column(DateTime(timezone=True), server_default=func.now(), nullable=False)
    updated_at: Mapped[DateTime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), onupdate=func.now(), nullable=False
    )
    is_deleted: Mapped[bool] = mapped_column(Boolean(), server_default="false", nullable=False)
    deleted_at: Mapped[DateTime | None] = mapped_column(DateTime(timezone=True), nullable=True)

    human_task: Mapped["HumanTask | None"] = relationship(  # noqa: F821
        "HumanTask", back_populates="model_query", uselist=False, cascade="all, delete-orphan"
    )
